snapcraft/project/_environment_checks.py

```diff
@@ -0,0 +1,132 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from typing import List, Optional  # noqa: F401
+from typing import Dict
+
+from snapcraft import project
+from snapcraft.internal import common, errors, os_release
+
+_MSG_NO_BASE_FOR_SNAP = (
+    "This snapcraft project does not specify the base keyword, "
+    "explicitly setting the base keyword enables the latest snapcraft features."
+)
+_MSG_BASE_FOR_DEB = (
+    "This version of snapcraft only has experimental support for the base keyword."
+)
+_MSG_UPGRADE = (
+    "Upgrade snapraft https://docs.snapcraft.io/t/upgrading-snapcraft/11658 to take "
+    "advantage of the latest features in snapcraft."
+)
+_MSG_HOST_MISMATCH_NO_BASE_TMPL = (
+    "This project is best built on '{expected_release_name} {expected_release_id}', "
+    "but is building on a '{release_name} {release_id}' host."
+)
+_MSG_HOST_MISMATCH_BASE_TMPL = (
+    "This project targets the {base!r} base, best built on "
+    "'{expected_release_name} {expected_release_id}', but is building on a "
+    "'{current_release_name} {current_release_id}' host."
+)
+_MSG_BASES_INFO = (
+    "Read more about bases at https://docs.snapcraft.io/t/base-snaps/11198"
+)
+
+
+class EnvironmentChecks:
+    def __init__(self, project: project.Project):
+        self._project = project
+        self._messages = None  # type: Optional[List[str]]
+
+    def get_messages(self) -> str:
+        # Unit tests are not setting the base, so we check for that here.
+        if self._project.info is None:
+            return
+        # If the message has been constructed, return it straight away.
+        if self._messages is not None:
+            return self._get_message_str()
+
+        self._messages = list()
+        self._check_base()
+        self._check_upgrade()
+        self._check_host()
+
+        # Lastly, if there were any messages we add our bases notification
+        # entry.
+        if self._messages:
+            self._messages.append(_MSG_BASES_INFO)
+
+        return self._get_message_str()
+
+    def _get_message_str(self) -> str:
+        return "\n".join(self._messages)
+
+    def _check_base(self) -> None:
+        base = self._project.info.base
+        if base is None and common.is_snap():
+            self._messages.append(_MSG_NO_BASE_FOR_SNAP)
+        elif base is not None and common.is_deb():
+            self._messages.append(_MSG_BASE_FOR_DEB)
+
+    def _check_upgrade(self) -> None:
+        # Any setup run from the debian package and not created from docker
+        # should get the upgrade message.
+        if not common.is_docker_instance() and common.is_deb():
+            self._messages.append(_MSG_UPGRADE)
+
+    def _check_host(self) -> None:
+        base = self._project.info.base
+        # core is not really a base, so it will not be set (legacy).
+        expected_release = self._project.get_expected_release_info_for_base(
+            base if base else "core"
```

Comment:
Would it perhaps be more intuitive to add:
`None: dict(codename="xenial", name="Ubuntu", version_id="16.04"),`
to _HOST_INFO_FOR_BASE?

Then I won't scratch my head wondering about the setting of base when there is no base? :)

Reply:
done